### PR TITLE
Fix some build issues, document some known issues

### PR DESCRIPTION
**Pull request by mchr3k**
## 🚀 Description
I used your template as the basis for exploring a best practices kotlin project - https://github.com/mchr3k/kotlin-project. In the process of creating this I hit some issues getting the template working in IntelliJ IDEA 2023.1.1 (Community Edition) which I've resolved in my fork of your template and which I'd now like to contribute back.

Changes:
* Fix gradle exception:
```
Execution failed for task ':buildSrc:compileKotlin'.
> 'compileJava' task (current target is 20) and 'compileKotlin' task (current target is 17) jvm target compatibility should be set to the same Java version.
  Consider using JVM toolchain: https://kotl.in/gradle/jvm/toolchain
```
by adding:
```
kotlin {
    jvmToolchain(17)
}
```
* Downgrade agp from 8.0.1 to 7.4.2 to address https://youtrack.jetbrains.com/issue/IDEA-317997. This was previously upgraded by renovate - https://github.com/cortinico/kotlin-android-template/commit/da28b85059fff345833d53957751ff6c3172f780.
* Remove `targetSdk = libs.versions.target.sdk.version.get().toInt()` which uses deprecated `targetSdk` attribute
* Remove `dokka` which appears to be unused since https://github.com/cortinico/kotlin-android-template/commit/3793a112a2ae5ed75a17cdf59a287c4bf5f13b21
* Fix detekt exception:
```
Execution failed for task ':app:detekt'.
> Invalid value (20) passed to --jvm-target, must be one of [1.6, 1.8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
```
by adding:
```
tasks {
    withType<Detekt>().configureEach {
        // Target version of the generated JVM bytecode. It is used for type resolution.
        jvmTarget = "17"
    }
}
```
* Integrate directly with ktlint rather than using `libs.detekt.formatting` which didn't appear to actually be checking code formatting as part of `./gradlew check`
* Upgrade org.jlleitschuh.gradle.ktlint to pickup the fix to this error:
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':library-kotlin:runKtlintFormatOverMainSourceSet'.
Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.KtLintWorkAction
Caused by: java.lang.ExceptionInInitializerError
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private transient java.lang.Object java.lang.Throwable.backtrace accessible: module java.base does not "opens java.lang" to unnamed module @2d72d8b1
```
* Add pre-commit hook to run `./gradlew check`

## 🧪 How Has This Been Tested?
I tested this by checking out https://github.com/mchr3k/kotlin-android-template locally onto my M1 Macbook Air running MacOS 13.3.1 (a) and opened the project in IntelliJ IDEA 2023.1.1 (Community Edition), using Gradle 8.1.1 for builds.

I made changes to the project until I could run `build` and `check` gradle tasks.

## 📦 Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

File: library-compose/build.gradle.kts
Comment (cortinico) [inline library-compose/build.gradle.kts:67]: This is unnecessary

File: library-android/build.gradle.kts
Comment (cortinico) [inline library-android/build.gradle.kts:66]: This is unnecessary

File: gradle/libs.versions.toml
Comment (cortinico) [inline gradle/libs.versions.toml:46]: I'm not sure we want to add this plugin here as I don't believe this is a industry standard in the Android ecosystem

File: gradle/libs.versions.toml
Comment (cortinico) [inline gradle/libs.versions.toml:2]: Please undo this. I understand that the project is not working in IDEA but that's a Google/Jetbrains problem which they should solve.

Once AGP hits stable, I'll update the template to match the latest stable as there is an Android Studio version out that can open it.

File: app/build.gradle.kts
Comment (cortinico) [inline app/build.gradle.kts:77]: This is unnecessary

File: README.md
Comment (cortinico) [inline README.md:75]: Removing Dokka is problematic as it breaks publishing to Maven Central so we should rather fix it instead of removing ti

File: build.gradle.kts
Comment (cortinico) [inline build.gradle.kts:26]: I'd rather prefer to stick with detekt formatting. What was wrong with it?

**Review by cortinico — CHANGES_REQUESTED** [review]
Thanks for sending this @mchr3k 🙏 really appreciate.

There are some points where I disagree with so it's maybe better to split it in smaller parts? Your call

File: README.md
Comment (mchr3k) [reply README.md:75]: I haven't worked with Dokka before, I just noticed that it's mentioned in the `libs.versions.toml` file but doesn't actually appear to be referenced anywhere in the Gradle build config. Does it get used implicitly somehow? https://github.com/Kotlin/dokka suggests that it should be mentioned in the gradle files somewhere.

So I think I've just removed Dokka as a dependency in `libs.versions.toml` and updated the README.md to reflect this change. I think Dokka was already not being used in this template. Have I missed something?

File: build.gradle.kts
Comment (mchr3k) [reply build.gradle.kts:26]: I made a change in one of the kotlin source files that is not consistent with the ktlint enforced code style and `./gradlew check` didn't detect this issue. Integrating directly with ktlint fixed this.

How do you confirm that ktlint is being applied to the kotlin source code through the detekt formatting plugin?

File: gradle/libs.versions.toml
Comment (mchr3k) [reply gradle/libs.versions.toml:2]: Do you mind if I revert this and replace it with a note in the README about this known issue?

File: library-compose/build.gradle.kts
Comment (mchr3k) [reply library-compose/build.gradle.kts:67]: Without this I get the following error:
```
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':library-compose:detekt'.
> Invalid value (20) passed to --jvm-target, must be one of [1.6, 1.8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
```

Comment (cortinico) [reply library-compose/build.gradle.kts:67]: That's because you're using JDK 20 which is unsupported with this template at the moment

File: gradle/libs.versions.toml
Comment (cortinico) [reply gradle/libs.versions.toml:46]: Nope there is no standard way to do so

File: gradle/libs.versions.toml
Comment (cortinico) [reply gradle/libs.versions.toml:2]: Yes please do. Also this is a recurrent issue of IntelliJ so not sure what we can do here actually

File: README.md
Comment (cortinico) [reply README.md:75]: Thanks for pointing out. We're fine removing dokka as we moved to `withJavadocJar()` some time ago and I forgot to clean this up.

File: build.gradle.kts
Comment (cortinico) [reply build.gradle.kts:26]: You're right that the formatting is not working fine. Please add this section at the bottom of `detekt.yml` and you will be formatting with the `check` step:

```
formatting:
  active: true
  android: false
  autoCorrect: true
  AnnotationOnSeparateLine:
    active: true
    autoCorrect: true
  AnnotationSpacing:
    active: true
    autoCorrect: true
  ArgumentListWrapping:
    active: true
    autoCorrect: true
    indentSize: 4
    maxLineLength: 120
  BlockCommentInitialStarAlignment:
    active: false
    autoCorrect: true
  ChainWrapping:
    active: true
    autoCorrect: true
  CommentSpacing:
    active: true
    autoCorrect: true
  CommentWrapping:
    active: false
    autoCorrect: true
    indentSize: 4
  DiscouragedCommentLocation:
    active: false
    autoCorrect: true
  EnumEntryNameCase:
    active: true
    autoCorrect: true
  Filename:
    active: true
  FinalNewline:
    active: true
    autoCorrect: true
    insertFinalNewLine: true
  FunKeywordSpacing:
    active: false
    autoCorrect: true
  FunctionReturnTypeSpacing:
    active: false
    autoCorrect: true
  FunctionSignature:
    active: false
    autoCorrect: true
    forceMultilineWhenParameterCountGreaterOrEqualThan: 2147483647
    functionBodyExpressionWrapping: 'default'
    maxLineLength: 120
    indentSize: 4
  FunctionStartOfBodySpacing:
    active: false
    autoCorrect: true
  FunctionTypeReferenceSpacing:
    active: false
    autoCorrect: true
  ImportOrdering:
    active: true
    autoCorrect: true
    layout: '*,java.**,javax.**,kotlin.**,^'
  Indentation:
    active: true
    autoCorrect: true
    indentSize: 4
  KdocWrapping:
    active: false
    autoCorrect: true
    indentSize: 4
  MaximumLineLength:
    active: true
    maxLineLength: 120
    ignoreBackTickedIdentifier: false
  ModifierListSpacing:
    active: false
    autoCorrect: true
  ModifierOrdering:
    active: true
    autoCorrect: true
  MultiLineIfElse:
    active: true
    autoCorrect: true
  NoBlankLineBeforeRbrace:
    active: true
    autoCorrect: true
  NoBlankLinesInChainedMethodCalls:
    active: true
    autoCorrect: true
  NoConsecutiveBlankLines:
    active: true
    autoCorrect: true
  NoEmptyClassBody:
    active: true
    autoCorrect: true
  NoEmptyFirstLineInMethodBlock:
    active: true
    autoCorrect: true
  NoLineBreakAfterElse:
    active: true
    autoCorrect: true
  NoLineBreakBeforeAssignment:
    active: true
    autoCorrect: true
  NoMultipleSpaces:
    active: true
    autoCorrect: true
  NoSemicolons:
    active: true
    autoCorrect: true
  NoTrailingSpaces:
    active: true
    autoCorrect: true
  NoUnitReturn:
    active: true
    autoCorrect: true
  NoUnusedImports:
    active: true
    autoCorrect: true
  NoWildcardImports:
    active: true
    packagesToUseImportOnDemandProperty: 'java.util.*,kotlinx.android.synthetic.**'
  NullableTypeSpacing:
    active: false
    autoCorrect: true
  PackageName:
    active: true
    autoCorrect: true
  ParameterListSpacing:
    active: false
    autoCorrect: true
  ParameterListWrapping:
    active: true
    autoCorrect: true
    maxLineLength: 120
  SpacingAroundAngleBrackets:
    active: true
    autoCorrect: true
  SpacingAroundColon:
    active: true
    autoCorrect: true
  SpacingAroundComma:
    active: true
    autoCorrect: true
  SpacingAroundCurly:
    active: true
    autoCorrect: true
  SpacingAroundDot:
    active: true
    autoCorrect: true
  SpacingAroundDoubleColon:
    active: true
    autoCorrect: true
  SpacingAroundKeyword:
    active: true
    autoCorrect: true
  SpacingAroundOperators:
    active: true
    autoCorrect: true
  SpacingAroundParens:
    active: true
    autoCorrect: true
  SpacingAroundRangeOperator:
    active: true
    autoCorrect: true
  SpacingAroundUnaryOperator:
    active: true
    autoCorrect: true
  SpacingBetweenDeclarationsWithAnnotations:
    active: true
    autoCorrect: true
  SpacingBetweenDeclarationsWithComments:
    active: true
    autoCorrect: true
  SpacingBetweenFunctionNameAndOpeningParenthesis:
    active: false
    autoCorrect: true
  StringTemplate:
    active: true
    autoCorrect: true
  TrailingCommaOnCallSite:
    active: false
    autoCorrect: true
    useTrailingCommaOnCallSite: true
  TrailingCommaOnDeclarationSite:
    active: false
    autoCorrect: true
    useTrailingCommaOnDeclarationSite: true
  TypeArgumentListSpacing:
    active: false
    autoCorrect: true
  TypeParameterListSpacing:
    active: false
    autoCorrect: true
  UnnecessaryParenthesesBeforeTrailingLambda:
    active: false
    autoCorrect: true
  Wrapping:
    active: true
    autoCorrect: true
    indentSize: 4
```

File: library-compose/build.gradle.kts
Comment (mchr3k) [reply library-compose/build.gradle.kts]: I've documented this in README.md. I had JDK 20 on my PATH which meant that even when I set JDK 17 within IDEA, Gradle was still using JDK 20.

File: gradle/libs.versions.toml
Comment (mchr3k) [reply gradle/libs.versions.toml]: Okay, I think including a pre-commit hook mechanism would be a good addition to this template, but I've removed it for now.

File: gradle/libs.versions.toml
Comment (mchr3k) [reply gradle/libs.versions.toml]: I've documented this in README.md

File: app/build.gradle.kts
Comment (mchr3k) [reply app/build.gradle.kts]: Fixed

File: build.gradle.kts
Comment (cortinico) [reply build.gradle.kts:21]: > How was this config derived?

That's from the official Detekt repo.
https://github.com/detekt/detekt/blob/main/detekt-formatting/src/main/resources/config/config.yml

or you can invoke `./gradlew detektGenerateConfig`

This whole point was already discussed here:
- https://github.com/cortinico/kotlin-android-template/pull/35

I'd rather keep detekt-formatting here as I'm one of the maintainer of Detekt, and you can do static analysis and formatting with just one gradle plugin rather than two. We've also surveyed the community and seems like a significant group of users are using this wrapper: 
- https://github.com/detekt/detekt/discussions/5997
So I'd rather keep it here for now.

File: README.md
Comment (cortinico) [inline README.md:47]: I'm unsure about this paragraph.

As you're using a jvmToolchain, Gradle will download and install the right version of the JDK which is needed for this project (17)

I believe there was a bug in AGP which resulted in the toolchain from being ignored, and should be fixed in AGP 8.1+

**Review by cortinico — CHANGES_REQUESTED** [review]

**Comment by cortinico** [comment]
Thanks for the follow-up @mchr3k, sorry for being late, but I was out in the last weeks. There are a couple of things we need to address, but we're good to go afterwards

File: README.md
Comment (mchr3k) [reply README.md:47]: I'm using Gradle 8.1.1 (checked by running `./gradlew --version` at the command line) and I get the error `Invalid value (20) passed to --jvm-target` if I have java 20 in my path and I don't get this error if I have java 17 in my path.

**Comment by mchr3k** [comment]
I've got one question left in this pull request around how to get detekt to apply formatting fixes. Once this has been resolved, I'll sort out the merge conflicts, and we can hopefully get this PR completed.

File: build.gradle.kts
Comment (mchr3k) [reply build.gradle.kts]: @cortinico - *bump* - I'd love to get an answer to this one last question so I can get this PR completed and merged.

Comment (cortinico) [reply build.gradle.kts]: Sorry for the late reply but I was (and I still am) quite busy at the moment so I'll be slower to respond. This PR is still in my inbox so I intend to to merge it 👍 

> Is there a gradle detekt task which will autofix formatting issues?

If `autoCorrect` is set to `true` in the Detekt config's file, detekt should run ktlint's autoCorrection on those issues

**Comment by cortinico** [comment]
> I've got one question left in this pull request around how to get detekt to apply formatting fixes. Once this has been resolved, I'll sort out the merge conflicts, and we can hopefully get this PR completed.

Answered (sorry for the late reply)

File: README.md
Comment (cortinico) [inline README.md:25]: Let's remove this

File: README.md
Comment (cortinico) [inline README.md:33]: I would also remove this section entirely as I don't think it's useful

File: README.md
Comment (cortinico) [reply README.md:47]: This is an issue with Detekt and not this template itself, so please remove

**Review by cortinico — COMMENTED** [review]
Hey @mchr3k I'd like to drive this to conclusion. I've left some comments. Please address them, otherwise I can also push to your branch and merge it

File: build.gradle.kts
Comment (cortinico) [reply build.gradle.kts]: > I'd love to get an answer to this one last question so I can get this PR completed and merged.

Following up here, if you run `./gradlew detekt --auto-correct` it will format those files

Comment (mchr3k) [reply build.gradle.kts]: Thanks for the response!

I've added some extra spaces on line 8 of this file:

```
package com.ncorti.kotlin.template.library

object FactorialCalculator {
    private const val MAX_FACTORIAL_64BIT = 20

    tailrec fun computeFactorial(input: Long, temp: Long = 1L): Long =
        when {
            input < 0 -> error(        "Factorial is not defined for negative numbers")
            input > MAX_FACTORIAL_64BIT -> error("Only a factorial up to 20 can fit in a 64-bit Long")
            input == 0L -> temp
            else -> computeFactorial(input - 1, temp * input)
        }
}
```

If I run `./gradlew detekt` I get the following errors:
```
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:32: Unnecessary long whitespace [NoMultipleSpaces]
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:31: Unexpected spacing after "(" [SpacingAroundParens]
```

This is despite `detekt.yml` specifying `autoCorrect: true` for both of these rules:
```
...
  NoMultipleSpaces:
    active: true
    autoCorrect: true
...
  SpacingAroundParens:
    active: true
    autoCorrect: true
...
```

If I run `./gradlew detekt --auto-correct` then the detekt task still fails:
```
> Task :library-kotlin:detekt FAILED
Property 'naming>BooleanPropertyNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'style>MandatoryBracesIfStatements' is deprecated. Use `BracesOnIfStatements` with `always` configuration instead.
Property 'style>ForbiddenComment>customMessage' is deprecated. Use `comments` and provide `reason` against each `value`..
Property 'style>ForbiddenComment>values' is deprecated. Use `comments` instead, make sure you escape your text for Regular Expressions..
Property 'naming>FunctionNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>FunctionParameterNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>ConstructorParameterNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>VariableNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:31: Unexpected spacing after "(" [SpacingAroundParens]
```

But the formatting issue has been fixed, and so `./gradlew detekt` now succeeds.

Questions:
* Why is the `--auto-correct` command line switch needed when `detekt.yml` specifies `autoCorrect: true`?
* Why does `Task :library-kotlin:detekt` fail when the only detected issue has been autocorrected? It looks to me like this is a bug caused by the same formatting issue triggering multiple rules and so the initial task failure outputs an error for `NoMultipleSpaces` and `SpacingAroundParens`, but the `--auto-correct` run only outputs an error related to `SpacingAroundParens`.

File: README.md
Comment (mchr3k) [reply README.md]: Fixed.

File: build.gradle.kts
Comment (cortinico) [reply build.gradle.kts]: The `--auto-correct` invocation will fail if the formatting is not valid. If you run it again, it should be true. The formatter needs a way to report if the formatting is valid or not, and that's what Detekt is using (build status).

> Why is the --auto-correct command line switch needed when detekt.yml specifies autoCorrect: true?

Because `autoCorrect: true` has granularity at the rule level. So you could enable/disable it for each rule or ruleset. But it won't actually trigger the autocorrection unless   the user specifies the `--auto-correct` flag.

File: README.md
Comment (mchr3k) [reply README.md:32]: When I was trying to get started with this template I was caught out by both of the issues I've listed here. I know these aren't technically issues with the template itself, but they are issues which someone using the template might trip over. How strongly do you feel about removing these? I think it would be helpful to leave them in.

File: README.md
Comment (mchr3k) [reply README.md:46]: This is an issue which I hit when trying to use this template and so I think it's helpful to have a workaround documented in the README. Of course you're the owner of this template so I'll remove this if you disagree but I want to double check - are you sure you want this to be removed?

File: build.gradle.kts
Comment (mchr3k) [reply build.gradle.kts]: Great, that makes sense, thanks! In that case I think this point is now resolved.

**Comment by mchr3k** [comment]
I've responded to all your comments. I think all that's left is for you to reply to the two outstanding points in the README.md diff to confirm whether or not you'd like me to remove the `Known issues` section. Once we have agreement about that this PR should be ready to be merged.

File: README.md
Comment (cortinico) [reply README.md:32]: How about we create a `KNOWN_ISSUES.md` file and we link it in the README?

**Comment by cortinico** [comment]
> I've responded to all your comments. I think all that's left is for you to reply to the two outstanding points in the README.md diff to confirm whether or not you'd like me to remove the `Known issues` section. Once we have agreement about that this PR should be ready to be merged.

I think we can move them to a separate .md file and link it here

File: README.md
Comment (mchr3k) [reply README.md]: Done

**Comment by mchr3k** [comment]
> > I've responded to all your comments. I think all that's left is for you to reply to the two outstanding points in the README.md diff to confirm whether or not you'd like me to remove the `Known issues` section. Once we have agreement about that this PR should be ready to be merged.
> 
> I think we can move them to a separate .md file and link it here.

Good idea! I've done that now which I hope means that this PR is now ready to be merged 🤞🏻

**Review by cortinico — APPROVED** [review]

**Comment by mchr3k** [comment]
Thanks for all your feedback and help with getting this change in good shape to be merged!

**Pull request merged**
